### PR TITLE
List - Optional Sync With Location

### PR DESCRIPTION
**Pull request by djhi**
Make the synchronization with the current location (URL) optional.

By default, this is enabled when a list is passed to a `<Resource>`. Otherwise the synchronization is opt-in using the `syncWithLocation` prop

- [x] Implementation
- [x] Tests

### CODE REVIEW

File: packages/ra-core/src/controller/useListParams.ts [hunk 1]
```diff
@@ -135,7 +139,9 @@ const useListParams = ({
         perPage,
     ];
 
-    const queryFromLocation = parseQueryFromLocation(location);
+    const queryFromLocation = syncWithLocation
+        ? parseQueryFromLocation(location)
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useListParams.ts:143]: the location is grabbed in `useListController` and passed to `useListParams`. I think we'd better call `useLocation` directly in `useListParams`.

File: packages/ra-core/src/controller/useListParams.ts [hunk 2]
```diff
@@ -161,13 +167,17 @@ const useListParams = ({
...
+                search: `?${stringify({
+                    ...newParams,
+                    filter: JSON.stringify(newParams.filter),
+                    displayedFilters: JSON.stringify(
+                        newParams.displayedFilters
+                    ),
+                })}`,
+            });
+        }
         dispatch(changeListParams(resource, newParams));
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useListParams.ts:181]: Not sure we should keep on relying on the Redux state when we don't sync with location

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Could you add unit tests?

File: packages/ra-core/src/controller/useListParams.ts (hunk 1 above)
Comment (djhi) [reply packages/ra-core/src/controller/useListParams.ts:143]: Done but not visible on this comment

File: packages/ra-core/src/controller/useListParams.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/useListParams.ts]: Done

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**